### PR TITLE
Extent setters in Cesium3DTileStyle

### PR DESCRIPTION
**Pull request by SunBlack**
See discussion [here](https://groups.google.com/forum/#!topic/cesium-dev/Kw14q6d1ODc)

### CODE REVIEW

**Comment by SunBlack** [comment]
No idea why tests are failing here, during tests in #5308 seems to pass currently. On my local machine tests currently stucks on 3d-tiles and my branch.

**Comment by pjcozzi** [comment]
I saw the forum discussion, and the approach is OK with me.  Thanks for the update, @SunBlack!

File: Source/Scene/Cesium3DTileStyle.js [hunk 1]
```diff
@@ -209,7 +209,7 @@ define([
         },
 
         /**
-         * Gets or sets the {@link StyleExpression} object used to evaluate the style's <code>show</code> property.
+         * Gets or sets the {@link StyleExpression} object used to evaluate the style's <code>show</code> property. Alternative an object defining a show style can be used.
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: Add examples of each of these cases to the doc.

File: Source/Scene/Cesium3DTileStyle.js [hunk 2]
```diff
@@ -246,13 +246,24 @@ define([
...
+                } else if (typeof value === 'string') {
+                    this._show = new Expression(value);
+                } else if (defined(value.conditions)) {
+                    this._show = new ConditionsExpression(value);
+                } else if (defined(value.expression) || defined(value.conditionsExpression)) {
+                    this._show = value;
+                } else {
+                    this._show = undefined;
+                    return;
+                }
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: Maybe it's not necessary to split up these two last cases. This would also allow someone to define their own evaluator like the example in the documentation.

```javascript
if (typeof value === 'boolean') {
    this._show = new Expression(String(value));
} else if (typeof value === 'string') {
    this._show = new Expression(value);
} else if (defined(value.conditions)) {
    this._show = new ConditionsExpression(value);
} else {
    this._show = value;
}
```

File: Source/Scene/Cesium3DTileStyle.js (hunk 1 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: Also tweak the new description a bit: `Alternatively a boolean, string, or object defining a show style can be used.`

File: Source/Scene/Cesium3DTileStyle.js [hunk 3]
```diff
@@ -246,13 +246,24 @@ define([
                 return this._show;
             },
             set : function(value) {
+                if (typeof value === 'boolean') {
+                    this._show = new Expression(String(value));
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: Now that `defines` are part of the styling language also pass those through to the expressions.

File: Source/Scene/Cesium3DTileStyle.js [hunk 4]
```diff
@@ -246,13 +246,24 @@ define([
...
+                } else if (defined(value.expression) || defined(value.conditionsExpression)) {
+                    this._show = value;
+                } else {
+                    this._show = undefined;
+                    return;
+                }
                 this._showShaderFunctionReady = false;
-                this._show = value;
             }
         },
```
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: The `setup` function can now be simplified be calling these setters within.

File: Source/Scene/Cesium3DTileStyle.js (hunk 3 above)
Comment (SunBlack) [inline Source/Scene/Cesium3DTileStyle.js]: I think this should be a separate PR, because we don't have a getter/setter for defines now. So I don't know if you really want old `defines` are still evaluated. And in case you allow getter and setter for `defines` you have to think other things, like if you need to recreate all expressions.

File: Source/Scene/Cesium3DTileStyle.js (hunk 4 above)
Comment (SunBlack) [inline Source/Scene/Cesium3DTileStyle.js]: Depends on discussion above ;)

File: Source/Scene/Cesium3DTileStyle.js (hunk 3 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: I don't think it hurts to pass the `defines` in even if they are no longer relevant to the new style, at the very least it does cut down on the code duplication. But if you would prefer that to be a different PR that is also ok with me.

File: Source/Scene/Cesium3DTileStyle.js (hunk 4 above)
Comment (SunBlack) [inline Source/Scene/Cesium3DTileStyle.js]: I having some trouble with a test, if I use setter in `setup`. In case I use new setter to initialize all variables following test fail: `return undefined shader functions when the style is empty`. But I think existing tests are wrong in this case. If e.g. no color is defined `DEFAULT_JSON_COLOR_EXPRESSION` will be used. So I don't understand why the test is expecting an undefined result instead of a shader generated from value `DEFAULT_JSON_COLOR_EXPRESSION`.

Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: For shader styling `undefined` is returned because otherwise `DEFAULT_JSON_COLOR_EXPRESSION` would cause all points to be white.

To fix this

```
that._colorShaderFunctionReady = !defined(styleJson.color);
that._showShaderFunctionReady = !defined(styleJson.show);
that._pointSizeShaderFunctionReady = !defined(styleJson.pointSize);
```

need to go below the setters in `setup`.

Comment (SunBlack) [inline Source/Scene/Cesium3DTileStyle.js]: It works, but I don't like code design here, because it is not really clear why one time `getColorShaderFunction` returns `undefined` and other times a valid shader, if both times value of `color` is equal to `DEFAULT_JSON_COLOR_EXPRESSION`.

So I would prefer: In case no color expression is defined `color` returns `undefined` instead of `DEFAULT_JSON_COLOR_EXPRESSION`. In `Cesium3DTileBatchTable.prototype.applyStyle` change code from:
```
            var color = style.color.evaluateColor(frameState, feature, scratchColor);
            var show = style.show.evaluate(frameState, feature);
```
to
```
            var color = defined(style.color) ? style.color.evaluateColor(frameState, feature, scratchColor) : 'white';
            var show = defined(style.show) ? style.show.evaluate(frameState, feature) : true;
```

Comment (lilleyse) [inline Source/Scene/Cesium3DTileStyle.js]: That approach seems good to me too.

Comment (SunBlack) [inline Source/Scene/Cesium3DTileStyle.js]: Behavior changed. I should be finished with the PR (hopefully all tests are passing on travis ;) )

**Comment by SunBlack** [comment]
No idea why the other tests are failing.

**Comment by lilleyse** [comment]
@SunBlack is this ready? The new changes look good to me.

**Comment by SunBlack** [comment]
Yes, i'm finished :)

**Comment by lilleyse** [comment]
Actually one small fix is needed in `Cesium3DTilesInspectorViewModel`:

`this._style.color.evaluateColor(frameState, currentFeature, scratchColor);` will need to check if `color` is defined first.

**Comment by SunBlack** [comment]
Maybe we should add a note to Readme about this changes. Any suggestion for a good short description?

**Comment by lilleyse** [comment]
A somewhat long description, but it works:

* Added ability to set a style's `color`, `show`, or `pointSize` with a string or object literal. `show` may also take a boolean and `pointSize` may take a number.

**Comment by SunBlack** [comment]
Should we mention that `Cesium3DTileStyle` don't is initialized anymore with default values, because this is a breaking change?

**Comment by lilleyse** [comment]
Yes also mention that in the breaking changes.

@pjcozzi would it be fine for this change to not go through deprecation? The change is that `color`, `show`, and `pointSize` properties of `Cesium3DTileStyle` can now be `undefined` rather than being set to default values.

**Comment by pjcozzi** [comment]
Maybe, I'll try to look at this soon.

**Comment by SunBlack** [comment]
In general I think it is okay to have here a breaking change, because 3d tiles are new to stable branch and I'm not sure if anyone is really using getter here. If not: We could add a deprecated warning to getter for release in 2 weeks and merge this MR direct after release. But I'm not sure if it is a good idea, because getter will be called at each tile and I'm not sure how I can filter it so message occurs only if it is really a deprecating case.

**Comment by pjcozzi** [comment]
Yes, this does not need to go through deprecation.

Please add a bit more doc to the set/get that explains that the `get` function doesn't always return exactly what is provided to the `set` function since it is converted to the native expression.

@lilleyse please merge when ready; I will not have the bandwidth to look at this again.

**Comment by lilleyse** [comment]
Thanks @SunBlack!

**Comment by SunBlack** [comment]
Tweaked doc is not fully correct, because if you set a custom object getter will return this custom object instead of an `Expression`, but i think it is okay ;)
